core/src/main/scala/kafka/server/DynamicBrokerConfig.scala

```diff
@@ -507,7 +507,7 @@ class DynamicBrokerConfig(private val kafkaConfig: KafkaConfig) extends Logging
 
   private def processReconfiguration(newProps: Map[String, String], validateOnly: Boolean): (KafkaConfig, List[BrokerReconfigurable]) = {
     val newConfig = new KafkaConfig(newProps.asJava, !validateOnly, None)
-    val updatedMap = updatedConfigs(newConfig.originalsFromThisConfig, currentConfig.originals)
+    val updatedMap = newConfig.originalsFromThisConfig.asScala
```

Comment:
Hmm... we don't want to re-apply configurations that haven't changed between the new and the old set of configs.  We want to generate a map that includes only the changes (whether they are additions or removals.)

Reply:
In the case of only deletions though, how would this work? For example, as in the test, if we first configured all the configs and then deleted them, my understanding was that we have to update the configs by applying all the configs in the original configs without the deleted configs.

Reply:
After thinking about it some more, I realize that the edge case that the test has is that if there are only delete operations done in a call to `incrementalAlterConfigs`, then `updatedConfigs` only returns an empty map, which then doesn't actually delete any of the configs. I've fixed this now by checking in `updatedConfigs` whether or not the new configs is a subset of the old configs, and if it is, simply return that as the `updatedMap`, if not, the old logic stands. I also wrote a new test that tests a mixture of operations (delete and set in one call). Please let me know what you think of this approach.

Reply:
There are other cases that need to be handled here.  For example, maybe some things are deleted and some things are added.  In that case, the approach in the PR now still doesn't work, since it falls back on the original incorrect logic.

The fundamental problem is that the `updatedMap` doesn't have entries for deleted elements, but there are parts of the code that equivocate between the updatedMap, and the set of all change deltas.  Either the `updatedMap needs to represent deleted elements somehow, or we need a separate data structure that does.

Reply:
I believe the problem with the previous logic was that it wasn't handling the edge case of when an `incrementalAlterConfig` operation has only `DELETE` operations. I did write a test when it deletes some configs but changes another config (in the same `incrementalAlterConfig` operation) in this PR: https://github.com/apache/kafka/pull/8260/files#diff-90c82107f654d6ec91c2d22c183b78c5R1728-R1766

The old logic just checks if the value of the old configs is not equal to the value of the new config, add that config to the `updatedMap`. I believe this covers all cases except the case of only `DELETE` operations, which is why I added a check to see if `newConfigs` were a subset of `currentConfigs`. If this is the case we can deduce that configs have only been deleted, so we only care about applying the `newConfigs` for the broker configuration. By applying the new configuration, we can effectively "delete" the configurations by reapplying all the new configs.

Reply:
Think about the case where some things are added and some things are removed.  In this case, your PR will fall back on the original incorrect logic.  As an example, let's say the old config is `(foo -> 123, bar -> 345)` and the new config is `(bar -> 345, baz -> 678)`.  The current logic in the PR will determine that the new config is not a subset of the old, and then generate an updated configuration that is `(foo -> 123, bar -> 345, baz -> 678)`.  Notice that foo is included even though it shouldn't be.

The reason why foo is included even though it shouldn't be is that `needsReconfiguration` uses the `updatedMap` to determine what needs to be reconfigured, and you never put foo in that map.

Reply:
I just modified the test I linked above to be the exact case you described, and it passes with the correct logic in the code. The test is as follows:

```scala
  @Test
  def testIncrementalAlterConfigsDeleteAndSetBrokerConfigs(): Unit = {
    client = Admin.create(createConfig())
    val broker0Resource = new ConfigResource(ConfigResource.Type.BROKER, "0")
// LeaderReplicationThrottledRateProp and FollowerReplicationThrottledRateProp are 
// foo and bar as described in your example
    client.incrementalAlterConfigs(Map(broker0Resource ->
      Seq(new AlterConfigOp(new ConfigEntry(DynamicConfig.Broker.LeaderReplicationThrottledRateProp, "123"),
          AlterConfigOp.OpType.SET),
        new AlterConfigOp(new ConfigEntry(DynamicConfig.Broker.FollowerReplicationThrottledRateProp, "456"),
          AlterConfigOp.OpType.SET)
      ).asJavaCollection).asJava).all().get()
    TestUtils.waitUntilTrue(() => {
      val broker0Configs = client.describeConfigs(Seq(broker0Resource).asJava).
        all().get().get(broker0Resource).entries().asScala.map {
        case entry => (entry.name, entry.value)
      }.toMap
      ("123".equals(broker0Configs.getOrElse(DynamicConfig.Broker.LeaderReplicationThrottledRateProp, "")) && "456".equals(broker0Configs.getOrElse(DynamicConfig.Broker.FollowerReplicationThrottledRateProp, "")))
    }, "Expected to see the broker properties we just set", pause=25)
// Now we delete LeaderReplicationThrottledRateProp and change 
// FollowerReplicationThrottledRateProp
// We also add ReplicaAlterLogDirsIoMaxBytesPerSecondProp, which would be baz as in your 
// example
    client.incrementalAlterConfigs(Map(broker0Resource ->
      Seq(new AlterConfigOp(new ConfigEntry(DynamicConfig.Broker.LeaderReplicationThrottledRateProp, ""),
        AlterConfigOp.OpType.DELETE),
        new AlterConfigOp(new ConfigEntry(DynamicConfig.Broker.FollowerReplicationThrottledRateProp, "654"),
          AlterConfigOp.OpType.SET),
        new AlterConfigOp(new ConfigEntry(DynamicConfig.Broker.ReplicaAlterLogDirsIoMaxBytesPerSecondProp, "987"),
          AlterConfigOp.OpType.SET)
      ).asJavaCollection).asJava).all().get()
    TestUtils.waitUntilTrue(() => {
      val broker0Configs = client.describeConfigs(Seq(broker0Resource).asJava).
        all().get().get(broker0Resource).entries().asScala.map {
        case entry => (entry.name, entry.value)
      }.toMap
      ("".equals(broker0Configs.getOrElse(DynamicConfig.Broker.LeaderReplicationThrottledRateProp, "")) &&
        "654".equals(broker0Configs.getOrElse(DynamicConfig.Broker.FollowerReplicationThrottledRateProp, "")) &&
        "987".equals(broker0Configs.getOrElse(DynamicConfig.Broker.ReplicaAlterLogDirsIoMaxBytesPerSecondProp, "")))
    }, "Expected to see the broker properties we just modified", pause=25)
  }
```

The logic for `updatedConfigs` isn't incorrect, it was missing out on the edge case of if an `incrementalAlterConfigs` request had only delete operations - that's why the check for subset is there, as if the new config is a subset of the old configs (where we can deduce the only alterations done were `DELETE` operations), then only the new configs should apply so that the old configs can be deleted. The old logic in question is here:

```scala
      newProps.asScala.filter {
        case (k, v) => v != currentProps.get(k)
      }
```

This logic filters the `newProps` on whether or not each key in `newProps` retrieved from `currentProps` has the same value as in `newProps`. Thus in the case you described, when the old config is `foo -> 123, bar -> 345)` and the new config is `(bar -> 345, baz -> 678)`, the logic will determine that the new config is not a subset of the old config, and generate an updated configuration by iterating through the keys of the new config.

There are only 2 keys in the new config it goes through, `bar` and `baz`. It checks the value of `bar` in the old config, and finds that the value of `bar` in the old config is not the same as the value in the new config, therefore this key-value pair is added. Similarly, it checks the value of `baz` in the old config, and since there is no entry for that key, it returns null, which is also not the same as the value of `baz` in the new config. Therefore that key-value pair is also added. Thus the `updatedMap` is returned as `(bar -> 345, baz -> 678)`, which is correct, as that is what we want applied.

However if we were filtering by iterating through `currentProps`, then we would run into the bug you mentioned.